eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelData.scala

```diff
@@ -303,30 +308,53 @@ case class LocalCommitPublished(commitTx: Transaction, claimMainDelayedOutputTx:
+      .collect { case txGen@TxGenerationResult.Success(tx) if confirmedTxs.contains(tx.input.outPoint.txid) => txGen }
+      .forall(txGen => txGen.isDone(irrevocablySpent))
+    isCommitTxConfirmed && isMainOutputConfirmed && allHtlcsSpent && allHtlcDelayedTxsSpent
+  }
+}
+
+object LocalCommitPublished {
+  /**
+   * We can always spend outgoing htlc outputs after the refund timeout expires, but for incoming htlcs it depends on
+   * what happens at the next hop. We track their status so that we know when we can consider a channel closed.
+   */
+  sealed trait HtlcOutputStatus
+  object HtlcOutputStatus {
+    /** We know how to spend this output */
+    case class Spendable(generationResult: TxGenerationResult[HtlcTx]) extends HtlcOutputStatus
+    /** We may be able to spend incoming htlcs outputs if we get the preimage from the next node */
+    case object PendingDownstreamSettlement extends HtlcOutputStatus
+    /** We know for sure that we will never be able to spend this incoming htlc output, because we got a failure from the next node in the route*/
+    case object Unspendable extends HtlcOutputStatus
   }
 }
 
 /**
  * Details about a force-close where they published their commitment.
  *
- * @param claimMainOutputTx tx claiming our main output (if we have one).
- * @param claimHtlcTxs      txs claiming HTLCs. There will be one entry for each pending HTLC. The value will be None
- *                          only for incoming HTLCs for which we don't have the preimage (we can't claim them yet).
- * @param claimAnchorTxs    txs spending anchor outputs to bump the feerate of the commitment tx (if applicable).
- *                          We currently only claim our local anchor, but it would be nice to claim both when it is
- *                          economical to do so to avoid polluting the utxo set.
+ * @param claimMainOutputTx_opt tx claiming our main output (if we have one).
+ * @param claimHtlcTxs          txs claiming HTLCs. There will be one entry for each pending HTLC. The value will be None
+ *                              only for incoming HTLCs for which we don't have the preimage (we can't claim them yet).
+ * @param claimAnchorTxs        txs spending anchor outputs to bump the feerate of the commitment tx (if applicable).
+ *                              We currently only claim our local anchor, but it would be nice to claim both when it is
+ *                              economical to do so to avoid polluting the utxo set.
  */
-case class RemoteCommitPublished(commitTx: Transaction, claimMainOutputTx: Option[ClaimRemoteCommitMainOutputTx], claimHtlcTxs: Map[OutPoint, Option[ClaimHtlcTx]], claimAnchorTxs: List[ClaimAnchorOutputTx], irrevocablySpent: Map[OutPoint, Transaction]) extends CommitPublished {
+case class RemoteCommitPublished(commitTx: Transaction,
+                                 claimMainOutputTx_opt: Option[TxGenerationResult[ClaimRemoteCommitMainOutputTx]],
```

Comment:
I see the distinction, which does make sense today for `option_saticremotekey` channels...but the future really is moving towards anchor outputs (and similar future constructs), so the pays-directly-to-wallet case is soon going to be legacy. So as you mention in your comment, it would probably make more sense to make that one of the `sealed trait` values instead of an `Option`, it should be more future-proof.